src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -656,70 +653,204 @@ IEnumerator IEnumerable.GetEnumerator()
                 {
+                    case '[':
+                    case '.':
+                        return true;
+                }
+            }
 
-                    if (key.StartsWith("[", StringComparison.OrdinalIgnoreCase))
-                    {
-                        var subKey = key.Substring(key.IndexOf('.') + 1);
+            return false;
+        }
 
-                        if (!subKey.StartsWith(prefix, StringComparison.OrdinalIgnoreCase))
-                        {
-                            continue;
-                        }
+        public PrefixEnumerable FindKeysWithPrefix(string prefix)
+        {
+            if (prefix == null)
+            {
+                throw new ArgumentNullException(nameof(prefix));
+            }
 
-                        if (string.Equals(prefix, subKey, StringComparison.Ordinal))
-                        {
-                            yield return entry;
-                            continue;
-                        }
+            return new PrefixEnumerable(this, prefix);
+        }
 
-                        key = subKey;
-                    }
-                    else
-                    {
-                        continue;
-                    }
+        public struct PrefixEnumerable : IEnumerable<KeyValuePair<string, ModelState>>
+        {
+            private readonly ModelStateDictionary _dictionary;
+            private readonly string _prefix;
+
+            public PrefixEnumerable(ModelStateDictionary dictionary, string prefix)
+            {
+                if (dictionary == null)
+                {
+                    throw new ArgumentNullException(nameof(dictionary));
                 }
 
-                // Everything is prefixed by the empty string
-                if (prefix.Length == 0)
+                if (prefix == null)
                 {
-                    yield return entry;
+                    throw new ArgumentNullException(nameof(prefix));
                 }
-                else
+
+                _dictionary = dictionary;
+                _prefix = prefix;
+            }
+
+            public PrefixEnumerator GetEnumerator()
+            {
+                return _dictionary == null ? new PrefixEnumerator() : new PrefixEnumerator(_dictionary, _prefix);
+            }
+
+            IEnumerator<KeyValuePair<string, ModelState>> IEnumerable<KeyValuePair<string, ModelState>>.GetEnumerator()
+            {
+                return GetEnumerator();
+            }
+
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                return GetEnumerator();
+            }
+        }
+
+        public struct PrefixEnumerator : IEnumerator<KeyValuePair<string, ModelState>>
+        {
+            private readonly ModelStateDictionary _dictionary;
+            private readonly string _prefix;
+
+            private bool _exactMatchUsed;
+            private IEnumerator<KeyValuePair<string, ModelState>> _enumerator;
```

Comment:
The other issue here is that this still allocates due to boxing. The changes in https://github.com/aspnet/Common/pull/53 will allow us to eliminate this change will still preserving the copy on write semantic.